### DISCUSSION TITLE
mono-repo strategy support?

### DISCUSSION THREAD

**Discussion by AdamJel** [pdm-project/pdm#1085] [Q&A]
Hello @frostming & all,

is there any way, how I can achieve the mono-repo strategy with pdm? That is, how can I have two independent (even mutually exclusive) sets of dependencies within one project?

In the world of `requirements.txt` that would mean to have two different reqs files - `requirements-alfa.txt` and `requirements-beta.txt` and {resolve dependency graph & compile lock file} for each req independently. 

The backstory for this is: with our team, we currently develop a new interface for our service. We decided to have just one repository for the easy of development workflow. In the meantime, we build two independent services - first with the old interface, second with the new interface. These two interfaces have significantly different dependencies. That is why we would like to be able to resolve dep-graph and compile lock file for each.

I know about optional deps `[project.optional-dependencies.<name>]` and about dev deps `[tool.pdm.dev-dependencies]`. But neither fulfil our needs, because the dependency graph is resolved across all these groups. E.g. the following hack attempt fails:

```bash
pdm add requests">2.2" --dev --group alfa
pdm add requests"<2.1" --dev --group beta
```

It is possible to compile into different lock files with the new `--lockfile` flag. But the input file (`pyproject.toml`) can still be just one (PEP defined, yes). So is there any other way how to achieve this?

Possible solution/implementation would be to allow to explicitly set the resolving for just some groups, e.g.:

```bash
pdm lock --lockfile "alfa.lock" --group alfa --isolated-resolving
```

This would be similar to [`environment`(s) in hatch](https://github.com/pypa/hatch/discussions/224),
or (hopefully soon to be implemented) [`subproject`(s) in poetry](https://github.com/python-poetry/poetry/issues/2270)

Thanks for any suggestions :)

**Comment by frostming** [comment]
Thanks for the suggestion, I think it makes much sense. Currently, we have a community-contributed plugin [Mina](https://github.com/GreyElaina/Mina) that can divide the project into different builds but it doesn't handle dependency resolution yet.

Maybe we can make improvements based on the project, the docs need to be improved as well.

**Reply by frostming** [reply]
Another way would be using the `--project/-p` to specify the project root, and the packages directory based on it.

**Reply by AdamJel** [reply]
If incorporating Mina plugin is possible and adding isolated resolving as well, that would be great. I have no insight into `pdm`'s code -> do you @frostming have any time estimate for the implementation? Thank you

**Reply by frostming** [reply]
> If incorporating Mina plugin is possible 

Not likely, as multi-project is a special use case and it is better to keep it as a plugin.

**Reply by AdamJel** [reply]
Ok, what is output of this discussion then? I understand, that you agree, that the described functionality is wanted. But I am not sure, what could be the next step to achieve it. Thanks

**Comment by GreyElaina** [comment]
uh...i maybe think of these commands, if all set, mina will support more pdm native commands' additional args.  
~~poor eng very sorry~~

**Reply by AdamJel** [reply]
You mean, you would implement the isolated resolving into the plugin?
